vendor/engines/bulk_email/spec/services/bulk_email/bulk_email_searcher_spec.rb

```diff
@@ -17,228 +17,346 @@
+          end
+        end
+      end
     end
 
-    it "returns just one product" do
-      params[:products] = [product.id]
-      expect(users).to eq([purchaser])
-      expect(searcher.order_details).to contain_all [@od1]
+    context "when filtering by reservation dates" do
+      let(:instrument) do
+        FactoryGirl.create(:instrument,
+                           facility_account: facility_account,
+                           min_reserve_mins: 60,
+                           max_reserve_mins: 60)
+      end
+
+      def place_reservation(purchaser, at)
+        place_reservation_for_instrument(purchaser, instrument, account, at)
+          .order_detail
+      end
+
+      let!(:od_yesterday) { place_reservation(purchaser, 1.day.ago) }
+      let!(:od_tomorrow) { place_reservation(purchaser2, 1.day.from_now) }
+      let!(:od_today) { place_reservation(purchaser3, Time.current) }
+
+      let(:start_date) { nil }
+      let(:end_date) { nil }
+
+      before(:each) do
+        params[:bulk_email][:start_date] = start_date
+        params[:bulk_email][:end_date] = end_date
+      end
+
+      it_behaves_like "active/inactive users" do
+        let(:user) { purchaser }
+      end
+
+      context "when the search start_date is today" do
+        let(:start_date) { usa_today }
+
+        context "and the end_date is unspecified" do
+          it "returns users who made reservations today and after" do
+            expect(users).to contain_exactly(purchaser3, purchaser2)
+          end
+        end
+
+        context "and the end_date is today" do
+          let(:end_date) { usa_today }
+
+          it "returns users who made reservations today only" do
+            expect(users).to eq [purchaser3]
+          end
+        end
+      end
+
+      context "when the search start_date is unspecified" do
+        context "and the end_date is today" do
+          let(:end_date) { usa_today }
+
+          it "returns users who made reservations today and before" do
+            expect(users).to contain_exactly(purchaser3, purchaser)
+          end
+        end
+      end
     end
 
-    it "returns two products" do
-      params[:products] = [product.id, product2.id]
-      expect(users).to contain_all [purchaser, purchaser2]
-      expect(searcher.order_details).to contain_all [@od1, @od2]
+    context "filtered by products" do
+      let!(:order_details) do
+        [
+          place_order(purchaser: purchaser, product: product, account: account),
+          place_order(purchaser: purchaser2, product: product2, account: account),
+          place_order(purchaser: purchaser3, product: product3, account: account),
+        ]
+      end
+
+      it_behaves_like "active/inactive users" do
+        let(:user) { purchaser }
+      end
+
+      context "when selecting no specific product" do
+        it "returns users who are customers of all products" do
+          expect(users).to contain_exactly(purchaser, purchaser2, purchaser3)
```

Comment:
I could see it being useful though, since we don't otherwise offer a blanket "anybody who purchased anything" option.